TorchProteinLibrary/FullAtomModel/Angles2Coords/Angles2Coords.py

```diff
@@ -29,60 +29,101 @@ def convertString(string):
 	return torch.from_numpy(np.fromstring(string+'\0', dtype=np.uint8))
 
+def convert2str(tensor):
+    return tensor.numpy().astype(dtype=np.uint8).tobytes().split(b'\00')[0]
+
 
 class Angles2CoordsFunction(Function):
 	"""
 	Protein angles -> coordinates function
 	"""
 	# @profile	
 	@staticmethod
-	def forward(ctx, input_angles_cpu, sequenceTensor, num_atoms):
-		ctx.save_for_backward(input_angles_cpu, sequenceTensor)
-		input_angles_cpu = input_angles_cpu.contiguous()
-		
-		max_num_atoms = torch.max(num_atoms)
-		batch_size = input_angles_cpu.size(0)
-		output_coords_cpu = torch.zeros(batch_size, 3*max_num_atoms, dtype=input_angles_cpu.dtype)
-		output_resnames_cpu = torch.zeros(batch_size, max_num_atoms, 4, dtype=torch.uint8)
-		output_atomnames_cpu = torch.zeros(batch_size, max_num_atoms, 4, dtype=torch.uint8)
-
-		_FullAtomModel.Angles2Coords_forward( 	sequenceTensor,
-												input_angles_cpu, 
-												output_coords_cpu, 
-												output_resnames_cpu,
-												output_atomnames_cpu)
-		
-		if math.isnan(output_coords_cpu.sum()):
-			raise(Exception('Angles2CoordsFunction: forward Nan'))
-		
-		return output_coords_cpu, output_resnames_cpu, output_atomnames_cpu, num_atoms
+	def forward(ctx, input_angles_cpu, sequenceTensor, num_atoms, polymer_type, chain_names):
+
+		if polymer_type == 0:
+			ctx.save_for_backward(input_angles_cpu, sequenceTensor, torch.tensor(polymer_type, dtype=torch.int32), chain_names)
+			input_angles_cpu = input_angles_cpu.contiguous()
+
+			max_num_atoms = torch.max(num_atoms)
+			batch_size = input_angles_cpu.size(0)
+			output_coords_cpu = torch.zeros(batch_size, 3*max_num_atoms, dtype=input_angles_cpu.dtype)
+			output_chainnames_cpu = torch.zeros(batch_size, max_num_atoms, 1, dtype=torch.uint8).fill_(ord('A'))
+			output_resnames_cpu = torch.zeros(batch_size, max_num_atoms, 4, dtype=torch.uint8)
+			output_resnums_cpu = torch.zeros(batch_size, max_num_atoms, dtype=torch.int)
+			output_atomnames_cpu = torch.zeros(batch_size, max_num_atoms, 4, dtype=torch.uint8)
+
+			_FullAtomModel.Angles2Coords_forward(sequenceTensor,
+												 input_angles_cpu,
+												 output_coords_cpu,
+												 output_resnames_cpu,
+												 output_resnums_cpu,
+												 output_atomnames_cpu,
+												 polymer_type,
+												 chain_names)
+
+			if math.isnan(output_coords_cpu.sum()):
+				raise(Exception('Angles2CoordsFunction: forward Nan'))
+
+			ctx.mark_non_differentiable(output_chainnames_cpu, output_resnames_cpu, output_resnums_cpu, output_atomnames_cpu, num_atoms)
+			return output_coords_cpu, output_chainnames_cpu, output_resnames_cpu, output_resnums_cpu, output_atomnames_cpu, num_atoms
+
+		if polymer_type == 1 or polymer_type == 2:
+			ctx.save_for_backward(input_angles_cpu, sequenceTensor, torch.tensor(polymer_type, dtype=torch.int32), chain_names)
+			input_angles_cpu = input_angles_cpu.contiguous()
+			max_num_atoms = num_atoms
```

Comment:
Why not "torch.max(num_atoms)" like for polymer_type 0? This code assumes "num_atoms" is a single number, not an array.